pkg/sync/sync_file.go

```diff
@@ -0,0 +1,703 @@
+	ctx, cancel := context.WithCancel(parentCtx)
+	sf := &SyncingFile{
+		lock: &sync.RWMutex{},
+		log: logrus.StandardLogger().WithFields(
+			logrus.Fields{
+				"component": "sync-file",
+				"filePath":  filePath,
+				"uuid":      uuid,
+			},
+		),
+		ctx:    ctx,
+		cancel: cancel,
+
+		filePath:         filePath,
+		tmpFilePath:      fmt.Sprintf("%s%s", filePath, TmpFileSuffix),
+		uuid:             uuid,
+		diskUUID:         diskUUID,
+		size:             size,
+		expectedChecksum: expectedChecksum,
+
+		state: types.StatePending,
+
+		handler: handler,
+	}
+	if size > 0 {
+		sf.log.WithField("size", size)
+	}
+
+	go func() {
+		// This may be time-consuming.
+		reuseErr := sf.checkAndReuseFile()
+		if reuseErr == nil {
+			return
+		}
+		sf.log.Infof("SyncingFile: cannot to reuse the existing file, will clean it up then start processing: %v", reuseErr)
+
+		var err error
+		defer func() {
+			sf.lock.Lock()
+			defer sf.lock.Unlock()
+			if err != nil {
+				sf.state = types.StateFailed
+				sf.log.Infof("SyncingFile: failed to init file syncing: %v", err)
+			} else {
+				sf.state = types.StateStarting
+				go sf.waitForProcessingBeginWithTimeout()
+			}
+		}()
+
+		dir := filepath.Dir(sf.filePath)
+		if mkErr := os.MkdirAll(dir, 0666); mkErr != nil && !os.IsExist(mkErr) {
+			err = mkErr
+			return
+		}
+		if err = os.RemoveAll(sf.tmpFilePath); err != nil {
+			return
+		}
+		if err = os.RemoveAll(sf.filePath); err != nil {
+			return
+		}
+	}()
+
+	return sf
+}
+
+func (sf *SyncingFile) checkAndReuseFile() (err error) {
+	sf.lock.RLock()
+	expectedChecksum := sf.expectedChecksum
+	filePath := sf.filePath
+	size := sf.size
+	sf.lock.RUnlock()
+
+	configFilePath := util.GetSyncingFileConfigFilePath(filePath)
+	defer func() {
+		if err == nil {
+			return
+		}
+		if rmErr := os.RemoveAll(configFilePath); rmErr != nil {
+			sf.log.Warnf("SyncingFile: failed to clean up config file %v after reuse failure: %v", configFilePath, rmErr)
+		}
+	}()
+
+	if expectedChecksum == "" {
+		return fmt.Errorf("cannot try to reuse a sync file when the expected checksum is unknown")
+	}
+
+	info, err := os.Stat(filePath)
+	if err != nil {
+		return err
+	}
+	if size > 0 && size != info.Size() {
+		return fmt.Errorf("sync file expected size %v doesn't match the existing file size %v", size, info.Size())
+	}
+
+	var currentChecksum string
+	config, err := util.ReadSyncingFileConfig(configFilePath)
+	if config != nil && config.ModificationTime == info.ModTime().UTC().String() {
+		logrus.Debugf("SyncingFile: directly get the checksum from a valid config during file reusage: %v", config.CurrentChecksum)
+		currentChecksum = config.CurrentChecksum
+	} else {
+		logrus.Debugf("SyncingFile: failed to get the checksum from a valid config during file reusage, will directly calculated it then")
+		currentChecksum, err = util.GetFileChecksum(filePath)
```

Comment:
Nit:
Just thinking that if we can open the file with buffered io rather than direct io.
The reason I ask is that the data prefetching mechanism of the filesystem may speed up the calculation of the checksum.

Reply:
Nice suggestion. I think YES.